pkg/kvstore/store/store.go

```
@@ -0,0 +1,395 @@
+
+	for _, key := range s.localKeys {
+		s.deleteLocalKey(key)
+	}
+}
+
+// keyPath returns the absolute kvstore path of a key
+func (s *SharedStore) keyPath(key Key) string {
+	// WARNING - STABLE API
+	return path.Join(s.conf.Prefix, key.GetKeyName())
+}
+
+// syncLocalKey synchronizes a key to the kvstore
+func (s *SharedStore) syncLocalKey(key LocalKey) error {
+	jsonValue, err := json.Marshal(key)
+	if err != nil {
+		return err
+	}
+
+	// Update key in kvstore, overwrite an eventual existing key, attach
+	// lease to expire entry when agent dies and never comes back up.
+	if err := kvstore.Update(s.keyPath(key), jsonValue, true); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// syncLocalKeys synchronizes all local keys with the kvstore
+func (s *SharedStore) syncLocalKeys() error {
+	s.mutex.RLock()
+	defer s.mutex.RUnlock()
+
+	for _, key := range s.localKeys {
+		if err := s.syncLocalKey(key); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// UpdateLocalKey adds a key to be synchronized with the kvstore
+func (s *SharedStore) UpdateLocalKey(key LocalKey) {
+	s.mutex.Lock()
+	s.localKeys[key.GetKeyName()] = key
+	s.mutex.Unlock()
+
+	key.OnDelete()
+}
+
+// UpdateLocalKeySync synchronously synchronizes a local key with the kvstore
+// and adds it to the list of local keys to be synchronized if the initial
+// synchronous synchronization was successful
+func (s *SharedStore) UpdateLocalKeySync(key LocalKey) error {
+	s.UpdateLocalKey(key)
+
+	if err := s.syncLocalKey(key); err != nil {
+		s.DeleteLocalKey(key)
+		return err
+	}
+
+	return nil
+}
+
+// deleteLocalKey must be called with s.mutex held
+func (s *SharedStore) deleteLocalKey(key LocalKey) {
+	err := kvstore.Delete(s.keyPath(key))
+
+	name := key.GetKeyName()
+	if _, ok := s.localKeys[name]; ok {
+		delete(s.localKeys, name)
+		key.OnDelete()
+
+		if err != nil {
+			s.getLogger().WithError(err).Warning("Unable to delete key in kvstore")
```

Is it intended to ignore the error if `localKey` is not present?